[WIP] Various improvements to folding

teto opened the pull request:
Follow up of https://github.com/neovim/neovim/issues/5931

- use the last column of fdc (vim doesn't for some reason)
- don't repeat symbol on wrapped lines
- show the end of a fold
- fillchars for fold symbols (see below)
- support unicode characters as fold markers
- improved style some places

I still have some things to do:
- support right/left display
- get rid of foldinfo_T ?
- also fillchar parsing is a bit weird, I mean the default value for each char needs to be entered twice, you first have a variable in globals.h then each char has its own dedicated function, for instance fillchar_vsep(int *attr) for fill_vert. I believe we shoudl change that.
For now if you want to see the characters, you have to explicitly set the characters with for instance:
```
if has("folding_fillchars")
  set fillchars+=foldopen:▾,foldsep:│,foldclose:▸,foldend:^
endif
```

Appearance of folds is not well tested (foldminlines either) but I intend to improve tests after this:
https://github.com/neovim/neovim/pull/5993

teto committed:
Various improvements to folding

- use the last column of fdc
- don't repeat symbol on wrapped lines
- fillchars for fold symbols
- support unicode characters as fold markers
- improved style some places

best used with:

if has("folding_fillchars")
  set fillchars+=foldopen:▾,foldsep:│,foldclose:▸,foldend:^
endif

teto committed:
fixed doc

teto committed:
mouse seems ok :D

justinmk commented on src/nvim/api/vim.c:
```diff
@@ -33,6 +34,22 @@
 # include "api/vim.c.generated.h"
 #endif
 
+/// @see foldCreate
```
would be better to move the doc here (and put `@see nvim_fold_foo` in the old place), since these doxygen comments get generated to `:help api-funcs`

teto replied:
Thanks for the tip ! I've still a few things to do before passing this to RFC.
I am now looking at making the fold width autoadapt. I enable it with `set fdc=-1` to prevent the creation of a new variable. Is that ok ? If the PR gets too big, I could split it up anyway.

justinmk replied:
Separate PRs would help a lot for this. Changing the folding code is fairly high-risk, so small PRs + testing is our best hope.

teto replied:
Tests are planned, that's why I started with https://github.com/neovim/neovim/pull/5993 .
If you could check it out that would be very cool.

teto committed:
wip

teto committed:
made many functions const related but

had to change the way the number of lines is updated

teto committed:
Pass window to foldCreate

- fdcwidth wrong on opening
- bug wwhen I zR a closed multifold, I have only open markers ('+')
- crash on exit :/

teto committed:
Fix cursor position for numbers

teto committed:
Fixed the fdc + number problem

The type < NOT_VALID condition prevented some variables from being
updated !

teto committed:
added fonctional/ui/fold_spec.lua test

...testing:
- display of the foldcolumn
- vim API nvim_fold_create/delete
- set foldminlines
- FoldColumn highlights

teto committed:
Updating tests

plexigras commented:
what is the status of this any chance this is gonna be a thing in the future?

teto commented:
I am waiting a bit to see how external marks and fold UIs can impact this part of the code. So for now I intend to focus first on the TUI cursor PRs. But I will update the PR as I do use it, I fancy the graphics very much.

plexigras commented:
i would love to have this, keep up the good work.

teto commented:
@plexigras I just rebased and after a quick test it seemed to work so feel free to try.

teto committed:
fixed older comments

teto commented:
considering I've not yet finished to merge my guicursor changes, this is here to stay for sure. Also because I expect folds might get externalized and screen.c improved so changes might necome  .

` I'd be in favor of hard-coding nicer chars`
while keeping the fillchars or should I remove those ? In case we keep both, we could have different defaults depending on ambiwidth as we did for the vertical separator

justinmk commented:
> while keeping the fillchars or should I remove those

You mean the new fillchar flags you added? No objection to new flags, but I was thinking this PR had stalled so hardcoding would be a nice quick improvement.

teto commented:
I've rebased and changed the default for ending a fold with yours. When using `set fillchars+=foldopen:▾,foldsep:│,foldclose:▸` it does look nice.
I noticed a bug when I set fdc=-1, the mouse won't work (at least for the one level case of folds, might be my PR).

justinmk commented on src/nvim/api/vim.c:
```diff
@@ -44,6 +45,45 @@
+/// @param start starting line
+/// @param end starting end
+/// @param err
+void nvim_fold_create(Window window, Integer start, Integer end, Error *err)
```
Should these be in `api/window.c` ?

see `:help dev-api` for API naming conventions: `nvim_win_set_fold()`,  `nvim_win_del_fold()`.

justinmk commented on src/nvim/buffer_defs.h:
```diff
@@ -1062,13 +1062,12 @@ struct window_S {
+  bool w_foldinvalid;       //!< when true: folding needs to be recomputed
+  int w_fdcwidth;           //!< optimal width to draw 'foldcolumn'
+  int w_nrwidth;            //!< width of 'number' and 'relativenumber'
+                            //!< column being used
```
I think doxygen expects three slashes, no? Otherwise the `!<` has no effect.

justinmk commented on src/nvim/eval.c:
```diff
@@ -10609,6 +10612,7 @@ static void f_has(typval_T *argvars, typval_T *rettv, FunPtr fptr)
     "find_in_path",
     "float",
     "folding",
+    "folding_fillchars",
```
We don't need these feature names since users can check `has('nvim-0.3')`. Also avoids having to guess what Vim will call a feature.

justinmk commented on src/nvim/fold.c:
```diff
@@ -1108,10 +1088,17 @@ static int foldLevelWin(win_T *wp, linenr_T lnum)
+    // int res = getDeepestNestingRecurse(&wp->w_folds);
+    // if(res != wp->w_fdcwidth) {
+    //   wp->w_fdcwidth = res;
+    // }
```
still needed?

justinmk commented on src/nvim/option.c:
```diff
@@ -4337,7 +4341,7 @@ static char *set_num_option(int opt_idx, char_u *varp, long value,
     }
   } else if (pp == &curwin->w_p_nuw) {
-    curwin->w_nrwidth_line_count = 0;
+    // curwin->w_nrwidth_line_count = 0;
```
?

justinmk commented on test/functional/legacy/045_folding_spec.lua:
```diff
@@ -213,5 +213,28 @@ describe('folding', function()
       line2]])
   end)
+
+  it('test fdc', function()
```
should go in the new test file rather than legacy.

teto committed:
appease linter

teto commented:
https://github.com/onivim/oni/issues/2086 prompts me to resume work on folds api. I am not sure what I can do but if we allow for fold externalization, shouldn't we give folds ids ?

KillTheMule commented:
Working on a plugin that does folding, I'd very much like an API for that. I'm not one for the large design questions, but here's what would be useful to me:

* Easy creation of multiple folds (right now, I running "{}, {}fo" though `nvim.command` several time)
* Easily find out which folds contain a given line (this is more important, since the user might add folds my addon knows nothing about).
* Deleting single folds without cursor movement.

Not saying this needs to be done, not sure if IDs help with all that, but just dropping my thoughts since I'm currently dealing with this :)

teto commented:
Some of it has already been merged. 
I am working towards the adaptive (auto) foldcolumn width. 
As for the API, I've moved some of it here https://github.com/neovim/neovim/pull/11523 but I am wary of defining an API without a consumer/usecase. I will close this issue soon.

teto commented on runtime/doc/fold.txt:
```diff
@@ -526,7 +526,8 @@ handles differently: Space, backslash and double-quote. |option-backslash|
-window to indicate folds.  When it is zero, there is no foldcolumn.  A normal
+window to indicate folds.  When it is -1, the width is found automatically.
+When it is zero, there is no foldcolumn.  A normal
 value is 4 or 5.  The minimal useful value is 2, although 1 still provides
```
I was not sure how to edit this paragraph, the mentions to '+' and '|' should now refer to fillchars foldclose/foldopen but this made the text a bit verbose.

teto commented:
To sum up, I've given up on `showing the end of a fold` since it required too many changes for a small gain.

The rest has been merged apart from the fold API, since I don't want to freeze an API I am an unsure about.
